store/tikv/gcworker/gc_worker.go

```diff
@@ -612,6 +670,84 @@ func doGCForOneRegion(bo *tikv.Backoffer, store tikv.Storage, safePoint uint64,
 	return nil, nil
 }
 
+func genNextGCTask(store tikv.Storage, bo *tikv.Backoffer, safePoint uint64, key kv.Key) (*gcTask, error) {
+	loc, err := store.GetRegionCache().LocateKey(bo, key)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	task := &gcTask{
+		startKey:  key,
+		endKey:    loc.EndKey,
+		safePoint: safePoint,
+	}
+	return task, nil
+}
+
+func doGC(ctx goctx.Context, store tikv.Storage, safePoint uint64, identifier string, concurrency int) error {
+	gcWorkerCounter.WithLabelValues("do_gc").Inc()
+
+	err := saveSafePoint(store.GetSafePointKV(), tikv.GcSavedSafePoint, safePoint)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Sleep to wait for all other tidb instances update their safepoint cache.
+	time.Sleep(gcSafePointCacheInterval)
+
+	log.Infof("[gc worker] %s start gc, concurrency %v, safePoint: %v.", identifier, concurrency, safePoint)
+	startTime := time.Now()
+	var successRegions int32
+	var failedRegions int32
+
+	ticker := time.NewTicker(gcJobLogTickInterval)
+	defer ticker.Stop()
+
+	// create task queue and task result queue, and start task workers
+	gcTaskCh := make(chan *gcTask, concurrency)
+	wg := sync.WaitGroup{}
+	for i := 0; i < concurrency; i++ {
+		w := newGCTaskWorker(store, gcTaskCh, &wg, identifier, &successRegions, &failedRegions)
+		wg.Add(1)
+		go w.run()
+	}
+
+	var key []byte
+	defer func() {
+		close(gcTaskCh)
+		wg.Wait()
+		log.Infof("[gc worker] %s finish gc, safePoint: %v, successful regions: %v, failed regions: %v, total cost time: %s",
+			identifier, safePoint, atomic.LoadInt32(&successRegions), atomic.LoadInt32(&failedRegions), time.Since(startTime))
+		gcHistogram.WithLabelValues("do_gc").Observe(time.Since(startTime).Seconds())
+	}()
+
+	bo := tikv.NewBackoffer(tikv.GcOneRegionMaxBackoff, ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.New("[gc worker] gc job canceled")
+		case <-ticker.C:
+			log.Infof("[gc worker] %s gc in process, safePoint: %v, successful regions: %v, failed regions: %v, total cost time: %s",
+				identifier, safePoint, atomic.LoadInt32(&successRegions), atomic.LoadInt32(&failedRegions), time.Since(startTime))
+		default:
+		}
+
+		task, err := genNextGCTask(store, bo, safePoint, key)
```

Comment:
where is the variable `key` initialized?

Reply:
what would happen when region split during different call of this function?

Reply:
no need to init, default values `""` is ok.

Reply:
task is defined by range, not region, so the worker will deal with the two regions if it splited.